Deferred Rendering Fixes

themadcreator opened the pull request:
Refactored deferred rendering in xyplot mostly into new `DeferredRenderer` class.

Added `postScale` projector to `.x()` and `.y()` setters to allow transformation in pixel coordinates while still storing the scale on the accessor binding.

Buffer canvas content while resizing to prevent janky redraws while resizing.

Make dataset changes use deferred rendering callback.

Exposed deferred render callback as `renderLowPriority`.

themadcreator committed:
Add deferreder renderer class

themadcreator committed:
Add postScale projector. Buffer canvas resizes

hellochar commented on src/plots/deferredRenderer.ts:
```diff
@@ -0,0 +1,105 @@
+ * `applyTransformCallback` again with an identity transform and finally invoke
+ * `renderCallback`, which should actually redraw the plot.
+ */
+export class DeferredRenderer<X, Y> {
```
💯

hellochar commented on src/plots/plot.ts:
```diff
@@ -157,14 +162,14 @@ export class Plot extends Component {
     this.animator(Plots.Animator.MAIN, mainAnimator);
     this.animator(Plots.Animator.RESET, new Animators.Null());
-    this._deferredResetEntityStore = Utils.Window.debounce(Plot._DEFERRED_RENDERING_DELAY, this._resetEntityStore);
+    this._deferredResetEntityStore = Utils.Window.debounce(DeferredRenderer.DEFERRED_RENDERING_DELAY, this._resetEntityStore);
```
this kinda hints that the entity store belongs in the renderer as well (later refactor?)

hellochar commented on src/plots/plot.ts:
```diff
@@ -199,6 +204,16 @@ export class Plot extends Component {
   public setBounds(width: number, height: number, originX?: number, originY?: number) {
     super.setBounds(width, height, originX, originY);
     if (this._canvas != null) {
+      if (this._bufferCanvas && !this._bufferCanvasValid) {
```
why do we need a buffer canvas?

themadcreator replied:
so, as soon as you change the size of a canvas with css or whatever, it clears the contents. we have some drag-resizable charts in various apps, and what was happening is that it would immediately do a full redraw while you drag-resize. This buffer canvas stores the current canvas contents when the resize starts and redraws it into the resized canvas. Eventually, the deferred rendering callback will trigger and do a full-rez redraw. If deferred rendering is disabled, the buffer copy will be overwritten immediately by a full redraw.

themadcreator replied:
Demonstration:
![update-after-resize](https://user-images.githubusercontent.com/569364/27448364-6a6e2ce0-5739-11e7-8ebf-0cf89f0ae4b2.gif)

hellochar reviewed (changes requested):
Awesome! can we confirm the bufferCanvas doesn't regress the window.devicePixelRatio bug?

themadcreator commented:
It uses the exact same size as the original canvas, so there should be no devicepixelratio issues.

themadcreator committed:
Fix unit test

Remove lint from default test commands since it's slow.
Add lint to circle.yml test commands.
Add big comment about buffer canvas.

Fix unit test by not resetting cachedDomain.

themadcreator committed:
webpack for tests

themadcreator committed:
Merge remote-tracking branch 'origin/develop' into bd/defer-scale-fix
